### PR TITLE
Custom layers with nGraph

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

wrap OpenCV layers which are not supported by Inference Engine backend to nGraph ops. It will let us to remove old API completely.

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.1.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/dnn.cpp [hunk 1]
```diff
@@ -1947,20 +1953,49 @@ struct Net::Impl
             Ptr<Layer> layer = ld.layerInstance;
             if (!fused && !layer->supportBackend(preferableBackend))
             {
-                addNgraphOutputs(ld);
-                net = Ptr<InfEngineNgraphNet>();
-                layer->preferableTarget = DNN_TARGET_CPU;
+                bool customizable = ld.id != 0 &&
+                                    INF_ENGINE_VER_MAJOR_GE(INF_ENGINE_RELEASE_2019R2) &&
```
Comment (l-bat) [inline modules/dnn/src/dnn.cpp]: Why do we need this condition? I mean nGraph appeared in 2020R1

File: modules/dnn/test/test_torch_importer.cpp [hunk 2]
```diff
@@ -359,11 +359,14 @@ TEST_P(Test_Torch_nets, ENet_accuracy)
...
         throw SkipTestException("");
     }
-#endif
-    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH && target != DNN_TARGET_CPU)
     {
-        applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
+        if (target == DNN_TARGET_OPENCL_FP16) applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_OPENCL_FP16, CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
+        if (target == DNN_TARGET_OPENCL)      applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_OPENCL, CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
+        if (target == DNN_TARGET_MYRIAD)      applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_MYRIAD, CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
```
Comment (dkurt) [inline modules/dnn/test/test_torch_importer.cpp]: @alalek, it looks like despite this flag, test is not skipped:
https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/3391
```
[ RUN      ] Test_Torch_nets.ENet_accuracy/3, where GetParam() = NGRAPH/MYRIAD
```
Shall I change it to MyriadX?

File: modules/dnn/src/ie_ngraph.cpp [hunk 3]
```diff
@@ -40,12 +69,81 @@ ngraphWrappers(const std::vector<Ptr<BackendWrapper> >& ptrs)
     return wrappers;
 }
 
+class NgraphCustomOp: public ngraph::op::Op {
+public:
+    static constexpr ngraph::NodeTypeInfo type_info{kOpenCVLayersType, 0};
+    const ngraph::NodeTypeInfo& get_type_info() const override { return type_info;  }
```
Comment (alalek) [inline modules/dnn/src/ie_ngraph.cpp]: I believe this should work too:
```
const ngraph::NodeTypeInfo& get_type_info() const override
{
    static constexpr ngraph::NodeTypeInfo type_info{kOpenCVLayersType, 0};
    return type_info;
}
```

File: modules/dnn/src/ie_ngraph.cpp [hunk 4]
```diff
@@ -40,12 +69,81 @@ ngraphWrappers(const std::vector<Ptr<BackendWrapper> >& ptrs)
...
+                                         std::vector<Mat>& outputs, std::vector<Mat>& internals)
+    : BackendNode(DNN_BACKEND_INFERENCE_ENGINE_NGRAPH), cvLayer(cvLayer_)
+{
+    std::ostringstream oss;
+    oss << (size_t)cvLayer.get();
+
+    std::map<std::string, InferenceEngine::Parameter> params = {
+        {"impl", oss.str()},
+        {"outputs", shapesToStr(outputs)},
+        {"internals", shapesToStr(internals)}
```
Comment (alalek) [inline modules/dnn/src/ie_ngraph.cpp]: Does it really necessary to serialize all of this into strings?
What is about related object lifetime handling?

File: modules/dnn/test/test_torch_importer.cpp [hunk 5]
```diff
@@ -351,6 +351,14 @@ TEST_P(Test_Torch_nets, ENet_accuracy)
 #if defined(INF_ENGINE_RELEASE) && INF_ENGINE_VER_MAJOR_EQ(2020010000)
     if (backend == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019)
         applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NN_BUILDER, CV_TEST_TAG_DNN_SKIP_IE_VERSION);
+
+    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH && target != DNN_TARGET_CPU)
```
Comment (alalek) [inline modules/dnn/test/test_torch_importer.cpp]: Perhaps this should be placed outside of this `#if` (see original location below)

File: modules/dnn/src/ie_ngraph.cpp (hunk 3 above)
Comment (dkurt) [reply modules/dnn/src/ie_ngraph.cpp]: Thanks! That works

File: modules/dnn/test/test_torch_importer.cpp (hunk 5 above)
Comment (dkurt) [reply modules/dnn/test/test_torch_importer.cpp]: Done, thanks!

File: modules/dnn/src/ie_ngraph.cpp (hunk 4 above)
Comment (dkurt) [reply modules/dnn/src/ie_ngraph.cpp]: I afraid not at current state of custom layers at Inference Engine. They still works with limited set of layers parameters: https://github.com/opencv/dldt/blob/b2140c0/inference-engine/src/inference_engine/ie_cnn_layer_builder.cpp#L82-L106

**Pull request merged**